tests/e2e/tests/pilot/pilot_test.go

```diff
@@ -17,155 +17,401 @@ package pilot
-		"Debug, skip clean up on failure")
-}
-
-func setup(env *tutil.Environment, t *testing.T) {
-	tutil.Tlog("Deploying infrastructure", spew.Sdump(env.Config))
-	if env.Err = env.Setup(); env.Err != nil {
-		t.Fatal(env.Err)
-	}
-}
-
-func teardown(env *tutil.Environment) {
-	env.Teardown()
-}
-
-func TestPilot(t *testing.T) {
-	if verbose {
-		config.Verbosity = 3
-	}
-
-	// Only run the tests if the user has defined the KUBECONFIG environment variable.
-	if config.KubeConfig == "" {
-		t.Skip("Env variable KUBECONFIG not set. Skipping tests")
-	}
-
-	if config.Hub == "" {
-		t.Skip("HUB not specified. Skipping tests")
-	}
-
-	if config.Tag == "" {
-		t.Skip("TAG not specified. Skipping tests")
-	}
-
-	env := tutil.NewEnvironment(*config)
-	defer teardown(env)
-	setup(env, t)
-
-	tests := []tutil.Test{
-		&http{Environment: env},
-		&grpc{Environment: env},
-		&tcp{Environment: env},
-		&headless{Environment: env},
-		&ingress{Environment: env},
-		&egressRules{Environment: env},
-		&routing{Environment: env},
-		&routingToEgress{Environment: env},
-		&zipkin{Environment: env},
-		&authExclusion{Environment: env},
-		&kubernetesExternalNameServices{Environment: env},
-		&authnPolicy{Environment: env},
-	}
-
-	for _, test := range tests {
-		// Run the test the configured number of times.
-		for i := 0; i < config.TestCount; i++ {
-			testName := test.String()
-
-			// User specified test doesn't match this test ... skip it.
-			if len(config.SelectedTest) > 0 && config.SelectedTest != testName {
-				t.Run(testName, func(t *testing.T) {
-					t.Skipf("Skipping test [%v] due to user-specified test: %v", t.Name(), config.SelectedTest)
-				})
-				continue
+func TestMain(m *testing.M) {
+	flag.Parse()
+	check(framework.InitLogging(), "cannot setup logging")
+	check(setTestConfig(), "could not create testConfig")
+	tc.Cleanup.RegisterCleanable(tc)
+	ret := tc.RunTest(m)
+	if ret != 0 {
+		// If any test fails, dump the logs for the k8s environment.
+		tc.Kube.DumpK8Env()
+	}
+	os.Exit(ret)
+}
+
+func setTestConfig() error {
+	var err error
+	tc, err = newTestConfig(*appHub, *appTag)
+	return err
+}
+
+func check(err error, msg string) {
+	if err != nil {
+		log.Errorf("%s. Error %s", msg, err)
+		os.Exit(-1)
+	}
+}
+
+// runRetriableTest runs the given test function the provided number of times.
+func runRetriableTest(t *testing.T, testName string, retries int, f func() error) {
+	t.Run(testName, func(t *testing.T) {
+		// Run all request tests in parallel.
+		// TODO(nmittler): Consider t.Parallel()?
+
+		remaining := retries
+		var errs error
+		for {
+			// Call the test function.
+			remaining--
+			err := f()
+			if err == nil {
+				// Test succeeded, we're done here.
+				return
 			}
 
-			if config.TestCount > 1 {
-				testName = testName + "_attempt_" + strconv.Itoa(i+1)
+			attempt := retries - remaining
+			errs = multierror.Append(errs, multierror.Prefix(err, fmt.Sprintf("attempt %d", attempt)))
```

Comment:
aren't you printing an increasingly larger errrs dump for every attempt? (like "attempt1", "attempt1, attempt2", ... ) ?

Reply:
+1 ... I've removed the log and just return the most recent error.